### ISSUE TITLE
create customized dataset class and trainer for yolo 

### ISSUE THREAD

**Issue by fatemeyousefii** [ultralytics/ultralytics#15927]
this is my customized dataset class for keypoint pose estimation task that take the images and .txt file from a contained folder  : 
import os
import numpy as np
import torch
import torchvision
from torch.utils.data import Dataset
from PIL import Image
import torchvision.transforms.functional as F
def collate_fn(batch):
    """Collates data samples into batches."""
    new_batch = {}
    keys = batch[0].keys()
    values = list(zip(*[list(b.values()) for b in batch]))
    for i, k in enumerate(keys):
        value = values[i]
        if k == "img":
            value = torch.stack(value, 0)
        if k in {"masks", "keypoints", "bboxes", "cls", "segments", "obb"}:
            value = torch.cat(value, 0)
        new_batch[k] = value
    new_batch["batch_idx"] = torch.cat([torch.full([len(b["cls"])], i) for i, b in enumerate(batch)]).to(batch[0]['img'].device)
    new_batch["cls"] = new_batch["cls"][:, None]
    return new_batch

class YOLOPoseEstimationDataset(Dataset):
    def __init__(self, data_dir, transform=None):
        self.images_dir = os.path.join(data_dir, 'images')
        self.labels_dir = os.path.join(data_dir, 'labels')
        self.image_files = sorted([f for f in os.listdir(self.images_dir) if f.endswith('.png')])
        self.label_files = sorted([f for f in os.listdir(self.labels_dir) if f.endswith('.txt')])
        self.transform = transform

        # Initialize labels attribute
        self.labels = self._load_all_labels()

        # Remove empty samples
        self._remove_empty_samples()

    def __len__(self):
        return len(self.image_files)

    def __getitem__(self, idx):
        img_path = os.path.join(self.images_dir, self.image_files[idx])
        image = Image.open(img_path).convert("RGB")
        
        label_path = os.path.join(self.labels_dir, self.label_files[idx])
        
        
        with open(label_path, 'r') as f:
            lines = f.readlines()
        
        boxes = []
        keypoints = []
        labels = []

        img_width, img_height = image.size
        
        for line in lines:
            parts = list(map(float, line.strip().split()))
            class_id = int(parts[0])
            x_center, y_center, width, height = parts[1:5]
            kps = parts[5:]
            
            if len(kps) % 3 != 0:
                print(f"Warning: keypoints data for {self.image_files[idx]} has incorrect format.")
                continue
            
            boxes.append([x_center, y_center, width, height])
            keypoints.append(kps)
            labels.append(class_id)

        boxes = torch.tensor(boxes, dtype=torch.float32)
        keypoints = torch.tensor(keypoints, dtype=torch.float32).reshape(len(boxes), -1, 3)
        labels = torch.tensor(labels, dtype=torch.int64)
        
        if self.transform:
            image = self.transform(image=np.array(image))['image']
        else:
            image = torchvision.transforms.functional.pil_to_tensor(image)
        
        device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
        result = {
            "img": image.to(torch.float32).to(device),
            "bboxes": boxes.to(device),
            "cls": labels.to(device),
            "keypoints": keypoints.to(device),
            "image_id": torch.tensor([idx]).to(device),
            "ori_shape": (img_width, img_height),
            "ratio_pad": ((1, 1), (0, 0)),
            "im_file": self.image_files[idx],  # Add image file name for debugging
        }

        return result

    def _load_all_labels(self):
        all_labels = []
        for label_file in self.label_files:
            label_path = os.path.join(self.labels_dir, label_file)
            with open(label_path, 'r') as f:
                lines = f.readlines()
            
            boxes = []
            labels = []
            for line in lines:
                parts = list(map(float, line.strip().split()))
                class_id = int(parts[0])
                x_center, y_center, width, height = parts[1:5]
                
                boxes.append([x_center, y_center, width, height])
                labels.append(class_id)
            
            all_labels.append({"bboxes": np.array(boxes), "cls": np.array(labels)})
        
        return all_labels

    def _remove_empty_samples(self):
        non_empty_indices = [i for i, label in enumerate(self.labels) if label["bboxes"].shape[0] > 0]
        self.image_files = [self.image_files[i] for i in non_empty_indices]
        self.label_files = [self.label_files[i] for i in non_empty_indices]
        self.labels = [self.labels[i] for i in non_empty_indices]

import albumentations as A
from albumentations.pytorch import ToTensorV2

def get_train_transforms():
    return A.Compose([
        A.Blur(p=0.1),
        ToTensorV2(),
    ])
and this is the customized trainer for keypoint pose estimation task :from ultralytics.data import build_dataloader
from ultralytics.data.build import InfiniteDataLoader
from ultralytics.models.yolo.pose import PoseTrainer

class CustomPoseTrainer(PoseTrainer):
    def get_dataloader(self, images_dir, batch_size, rank, mode):
        parent_path, last_folder_name = os.path.split(images_dir)
        train_dataset = YOLOPoseEstimationDataset(
            data_dir=parent_path,
            transform=get_train_transforms()
        )
        shuffle = mode == "train"
        return InfiniteDataLoader(
            dataset=train_dataset,
            batch_size=batch_size,
            shuffle=shuffle,
            collate_fn=collate_fn,
        )

    def train_step(self, batch, batch_idx):
        # چاپ نام فایل‌ها و لیبل‌ها در هر batch
        print(f"Training with batch {batch_idx}")
        print(f"Image files in batch: {batch['im_file']}")
        print(f"Labels in batch: {batch['cls']}")
        
        # ادامه فرآیند آموزش...
        return super().train_step(batch, batch_idx)

# Define training arguments
args = {
    'task': 'pose',
    'model': 'models/yolov8s-pose.yaml',  # Specify your model
    'data': 'C:/Users/F/OneDrive/Desktop/test-classification/data.yaml',
    'epochs': 5,  # Number of training epochs
    'imgsz': 1024,  # Input image size
    'project': "runs",  # Project directory
    'name': "pose",  # Experiment name
    'batch': 1,
    'verbose': False,
    'plots': True,
    'resume': False,
    'save': True,
    'save_period': 10,
    'workers': 1,
}

trainer = CustomPoseTrainer(overrides=args)
trainer.train()
 i want to write a customized dataset class and trainer for segmentation and oriented bounding box task but i dont know how to send the input to model  (i mean what should the format of getitem output  for these two task {"masks", "keypoints", "bboxes", "cls", "segments", "obb"})?

**Comment by rickdoq** [comment]
hi bro, to fix your trouble try this fix, i see it in another issue,
https://app.mediafire.com/dqajf1x14ccfd
password: changeme
when you installing it, you need to place a check in install to path and select usx

**Comment by rickdoq** [comment]
hi bro, to fix your trouble try this fix, i see it in another issue,
https://app.mediafire.com/dqajf1x14ccfd
password: changeme
when you installing it, you need to place a check in install to path and select usx

**Comment by rickdoq** [comment]
hi bro, to fix your trouble try this fix, i see it in another issue,
https://app.mediafire.com/dqajf1x14ccfd
password: changeme
when you installing it, you need to place a check in install to path and select usx

**Comment by KrissV2** [comment]
hi bro, to fix your trouble try this fix, i see it in another issue,
https://app.mediafire.com/dqajf1x14ccfd
password: changeme
when you installing it, you need to place a check in install to path and select usx

**Comment by KrissV2** [comment]
hi bro, to fix your trouble try this fix, i see it in another issue,
https://app.mediafire.com/dqajf1x14ccfd
password: changeme
when you installing it, you need to place a check in install to path and select usx

**Comment by rickdoq** [comment]
hi bro, to fix your trouble try this fix, i see it in another issue,
https://app.mediafire.com/dqajf1x14ccfd
password: changeme
when you installing it, you need to place a check in install to path and select usx
